File: python/gink/impl/pair_set.py
""" Contains the pair set class definition """

from typing import Optional, Tuple, Iterable, Union, Set
from typeguard import typechecked
from .database import Database
from .muid import Muid
from .container import Container
from .coding import PAIR_SET, deletion, inclusion
from .bundler import Bundler
from .typedefs import GenericTimestamp
from .utilities import normalize_pair, experimental

Pair = Tuple[Union[Container, Muid], Union[Container, Muid]]

@experimental
class PairSet(Container):
    _missing = object()
    _BEHAVIOR = PAIR_SET

    @typechecked
    def __init__(
                self,
                *,
                muid: Optional[Union[Muid, str]] = None,
                contents: Optional[Iterable[Pair]] = None,
                database: Optional[Database] = None,
                bundler: Optional[Bundler] = None,
                comment: Optional[str] = None,
            ):
        """
        Constructor for a pair set proxy.

        muid: the global id of this container, created on the fly if None
        contents: optionally expecting a dictionary of {"include": Set, "exclude": Set} to prefill the pair set
        database: database send bundles through, or last db instance created if None
        bundler: the bundler to add changes to, or a new one if None and immediately commits
        comment: optional comment to add to the bundler
        """
        database = database or Database.get_most_recently_created_database()
        immediate = False
        if bundler is None:
            immediate = True
            bundler = database.bundler(comment)
        created = False
        if isinstance(muid, str):
            muid = Muid.from_str(muid)
        elif muid is None:
            muid = Container._create(PAIR_SET, bundler=bundler)
            created = True
        Container.__init__(self, muid=muid, database=database)
        if contents:
            assert isinstance(contents, dict)
            assert contents.keys() <= {"include", "exclude"}, "expecting only 'include' and 'exclude' keys in contents"
            if not created:
                self.clear(bundler=bundler)
            included = contents.get("include", set())
            assert isinstance(included, Iterable)
            for pair in included:
                assert isinstance(pair, tuple) and len(pair) == 2
                self.include(pair, bundler=bundler) # type: ignore

            excluded = contents.get("exclude", set())
            assert isinstance(excluded, Iterable)
            for pair in excluded:
                assert isinstance(pair, tuple) and len(pair) == 2
                self.exclude(pair, bundler=bundler) # type: ignore

        if immediate and len(bundler):
            bundler.commit()

    @typechecked
    def include(self, pair: Pair, *, bundler: Optional[Bundler]=None, comment: Optional[str]=None):
        """ Includes a pair of Containers or Muids in the pair set """
        return self._add_entry(key=pair, value=inclusion, bundler=bundler, comment=comment)

    @typechecked
    def exclude(self, pair: Pair, *, bundler: Optional[Bundler]=None, comment: Optional[str]=None):
        """ Excludes a pair of Containers or Muids from the pair set """
        return self._add_entry(key=pair, value=deletion, bundler=bundler, comment=comment)

    @typechecked
    def contains(self, pair: Pair, *, as_of: GenericTimestamp = None) -> bool:
        """ Returns True if the pair is included the pair set """
        ts = self._database.resolve_timestamp(as_of)
        pair = normalize_pair(pair)
        found = self._database.get_store().get_entry_by_key(self.get_muid(), key=pair, as_of=ts)
        return bool(found and not found.builder.deletion)

    @typechecked
    def __contains__(self, pair: Pair) -> bool:
        return self.contains(pair)

    def get_pairs(self, *, as_of: GenericTimestamp = None) -> Set[Tuple[Muid, Muid]]:
        """ Returns a set of muid pairs included in the pair set at a given time """
        as_of = self._database.resolve_timestamp(as_of)
        iterable = self._database.get_store().get_keyed_entries(
            container=self._muid, as_of=as_of, behavior=self._BEHAVIOR)

        return {(Muid.create(builder=entry_pair.builder.pair.left, context=entry_pair.address),
                Muid.create(builder=entry_pair.builder.pair.rite, context=entry_pair.address))
                for entry_pair in iterable if not entry_pair.builder.deletion}

    def __iter__(self) -> Iterable[Tuple[Muid, Muid]]:
        for pair in self.get_pairs():
            yield pair

    def size(self, *, as_of: GenericTimestamp = None) -> int:
        """ Returns the number of elements included in the pair set, optionally at a given time """
        ts = self._database.resolve_timestamp(as_of)
        iterable = self._database.get_store().get_keyed_entries(
            container=self._muid, as_of=ts, behavior=self._BEHAVIOR)
        count = 0
        for entry_pair in iterable:
            if not entry_pair.builder.deletion:
                count += 1
        return count

    def dumps(self, as_of: GenericTimestamp = None) -> str:
        """ Returns the contents of this container as a string """
        as_of = self._database.resolve_timestamp(as_of)
        identifier = f"muid={self._muid!r}"
        result = f"""{self.__class__.__name__}({identifier}, contents="""
        result += "{"

        included_stuffing = "'include': [\n\t"
        excluded_stuffing = "'exclude': [\n\t"
        something = False
        for entry_pair in self._database.get_store().get_keyed_entries(
            container=self.get_muid(), behavior=self._BEHAVIOR, as_of=as_of):
            something = True
            left = Muid.create(builder=entry_pair.builder.pair.left)
            rite = Muid.create(builder=entry_pair.builder.pair.rite)
            if entry_pair.builder.deletion:
                excluded_stuffing += f"({left!r}, {rite!r}),\n\t"
            else:
                included_stuffing += f"({left!r}, {rite!r}),\n\t"

        if something:
            result += "\n\t"
        if included_stuffing != "'include': [\n\t":
            result += "".join(included_stuffing) + "],"
        if excluded_stuffing != "'exclude': [\n\t":
            result += "".join(excluded_stuffing) + "],"

        result += "})"
        return result

File: javascript/implementation/PairSet.ts
import { Database } from "./Database";
import { Container } from "./Container";
import { Muid, AsOf, MuidTuple } from "./typedefs";
import { Bundler } from "./Bundler";
import { ensure, muidToString, fromStorageKey } from "./utils";
import { toJson } from "./factories";
import { Behavior, ContainerBuilder } from "./builders";

export class PairSet extends Container {
    constructor(
        database: Database,
        address: Muid,
        containerBuilder?: ContainerBuilder
    ) {
        super(database, address, Behavior.PAIR_SET);
        if (this.address.timestamp < 0) {
            ensure(address.offset === Behavior.PAIR_SET);
        } else {
            ensure(containerBuilder.getBehavior() === Behavior.PAIR_SET);
        }
    }

    /**
     * Includes a pair of Muids or Containers in the pair set.
     * @param key a pair of either containers or Muids to include
     * @param change an optional bundler to put this change into
     * @returns a promise that resolves to the Muid for the inclusion
     */
    async include(
        key: [Container, Container],
        change?: Bundler | string
    ): Promise<Muid> {
        return await this.addEntry(key, Container.INCLUSION, change);
    }

    /**
     * Excludes a pair of Muids or Containers in the pair set.
     * @param key a pair of either containers or Muids to include
     * @param change an optional bundler to put this change into
     * @returns a promise that resolves to the Muid for the exclusion
     */
    async exclude(
        key: [Container, Container],
        change?: Bundler | string
    ): Promise<Muid> {
        return await this.addEntry(key, Container.DELETION, change);
    }

    /**
     * If the pair set has a key or not.
     * @param key array of 2 muids or containers
     * @param asOf optional timestamp to look back to
     * @returns a promise that resolves to a boolean, true if the key is included, false if not
     */
    async contains(
        key: [Muid | Container, Muid | Container],
        asOf?: AsOf
    ): Promise<boolean> {
        const aKey: [Muid, Muid] = [
            key[0] instanceof Container ? key[0].address : key[0],
            key[1] instanceof Container ? key[1].address : key[1],
        ];
        const found = await this.database.store.getEntryByKey(
            this.address,
            aKey,
            asOf
        );
        if (found && found.deletion) return false;
        return Boolean(found);
    }

    /**
     *
     * @param args Optional arguments, including:
     * @argument toTime Optional time to reset to. If absent, the container will be cleared.
     * @argument bundlerOrComment Optional bundler or comment to add this change to
     * @argument skipProperties If true, do not reset properties of this container. By default,
     * all properties associated with this container will be reset to the time specified in toTime.
     * @argument recurse NOTE: THIS FLAG IS IGNORED. Recursive reset for Inclusion-based containers
     * is not yet implemented, but this arg needs to be accepted for other containers recursively
     * resetting this one.
     * @argument seen NOTE: THIS FLAG IS IGNORED. Recursive reset for Inclusion-based containers
     * is not yet implemented, but this arg needs to be accepted for other containers recursively
     * resetting this one.
     */
    async reset(args?: {
        toTime?: AsOf;
        bundlerOrComment?: Bundler | string;
        skipProperties?: boolean;
        recurse?: boolean;
        seen?: Set<string>;
    }): Promise<void> {
        const toTime = args?.toTime;
        const bundlerOrComment = args?.bundlerOrComment;
        const skipProperties = args?.skipProperties;
        let immediate = false;
        let bundler: Bundler;
        if (bundlerOrComment instanceof Bundler) {
            bundler = bundlerOrComment;
        } else {
            immediate = true;
            bundler = new Bundler(bundlerOrComment);
        }
        if (!toTime) {
            // If no time is specified, we are resetting to epoch, which is just a clear
            this.clear(false, bundler);
        } else {
            const union = new Set<[MuidTuple, MuidTuple]>();
            const entriesThen = await this.database.store.getKeyedEntries(
                this.address,
                toTime
            );
            const entriesNow = await this.database.store.getKeyedEntries(
                this.address
            );

            for (const [key, entry] of entriesThen) {
                const storageKey = <[MuidTuple, MuidTuple]>entry.storageKey;
                union.add(storageKey);
            }
            for (const [key, entry] of entriesNow) {
                const storageKey = <[MuidTuple, MuidTuple]>entry.storageKey;
                union.add(storageKey);
            }
            for (const key of union) {
                const genericKey = fromStorageKey(key);
                const thenEntry = await this.database.store.getEntryByKey(
                    this.address,
                    genericKey,
                    toTime
                );
                const nowEntry = await this.database.store.getEntryByKey(
                    this.address,
                    genericKey
                );
                ensure(nowEntry || thenEntry, "both then and now undefined?");
                if (!nowEntry) {
                    // This key was present then, but not now, so we need to add it back
                    ensure(thenEntry, "missing then entry?");
                    await this.addEntry(genericKey, thenEntry.value, bundler);
                } else if (!thenEntry) {
                    // This key is present now, but not then, so we need to delete it
                    ensure(nowEntry, "missing now entry?");
                    await this.addEntry(
                        genericKey,
                        Container.DELETION,
                        bundler
                    );
                } else if (nowEntry.deletion !== thenEntry.deletion) {
                    if (nowEntry.deletion) {
                        // Present then, deleted now. Need to revive.
                        await this.addEntry(
                            genericKey,
                            Container.INCLUSION,
                            bundler
                        );
                    } else if (thenEntry.deletion) {
                        // Present now, deleted then. Need to delete.
                        await this.addEntry(
                            genericKey,
                            Container.DELETION,
                            bundler
                        );
                    }
                } else {
                    ensure(
                        nowEntry.deletion === thenEntry.deletion,
                        "last case should be same entry"
                    );
                }
            }
        }
        if (!skipProperties) {
            await this.database.resetContainerProperties(this, toTime, bundler);
        }
        if (immediate) {
            await this.database.addBundler(bundler);
        }
    }

    /**
     * The number of items in the pair set.
     * @param asOf optional timestamp to look back to
     * @returns a promise that resolves to the number of entries
     */
    async size(asOf?: AsOf): Promise<number> {
        const entries = await this.database.store.getKeyedEntries(
            this.address,
            asOf
        );
        return entries.size;
    }

    /**
     * All of the pairs in the Pair Set as a set
     * @param asOf optional timestamp to look back to
     * @returns a promise that resolves to a set of pairs [Muid, Muid]
     */
    async getPairs(asOf?: AsOf): Promise<Set<Array<Muid>>> {
        const entries = await this.database.store.getKeyedEntries(
            this.address,
            asOf
        );
        const toSet = new Set<Array<Muid>>();
        for (const [key, entry] of entries) {
            if (!entry.deletion) {
                toSet.add(<Array<Muid>>entry.storageKey);
            }
        }
        return toSet;
    }

    /**
     * Generates a JSON representation of the data in the pair set.
     * Mostly intended for demo/debug purposes.
     * @param indent true to pretty print (not yet implemented)
     * @param asOf optional timestamp to look back to
     * @param seen (internal use only! This prevents cycles from breaking things)
     * @returns a JSON string
     */
    async toJson(
        indent: number | boolean = false,
        asOf?: AsOf,
        seen?: Set<string>
    ): Promise<string> {
        //TODO(https://github.com/google/gink/issues/62): add indentation
        ensure(indent === false, "indent not implemented");
        if (seen === undefined) seen = new Set();
        const mySig = muidToString(this.address);
        if (seen.has(mySig)) return "null";
        seen.add(mySig);
        const asSet = await this.getPairs(asOf);
        let returning = "[";
        let first = true;
        for (const key of asSet) {
            if (first) {
                first = false;
            } else {
                returning += ",";
            }
            returning += await toJson(
                `[${muidToString(key[0])}, ${muidToString(key[1])}]`,
                indent === false ? false : +indent + 1,
                asOf,
                seen
            );
        }
        returning += "]";
        return returning;
    }
}
